src/libsyntax/parse/parser.rs

```diff
@@ -635,13 +638,64 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn recover_closing_delimiter(
+        &mut self,
+        tokens: &[token::Token],
+        mut err: DiagnosticBuilder<'a>,
+    ) -> PResult<'a, ()> {
+        let mut pos = None;
+        let mut tokens: Vec<token::Token> = tokens.to_vec();
+        tokens.extend(self.expected_tokens.iter().filter_map(|t| match t {
+            TokenType::Token(t) => Some((*t).clone()),
+            _ => None,
+        }));
+        // we want to use the last closing delim that would apply
+        for (i, (delim, span)) in self.open_braces.iter().enumerate().rev() {
+            if tokens.contains(&token::CloseDelim(*delim)) && self.span > *span {
+                pos = Some(i);
+                break;
+            }
+        }
+        match pos {
+            Some(pos) => {
+                // Recover and assume that the detected unclosed delimiter was meant for
+                // this location. Emit the diagnostic and act as if the delimiter was
+                // present for the parser's sake.
+
+                // Don't attempt to recover from this unclosed delimiter more than once.
+                let (delim, open_sp) = self.open_braces.remove(pos);
+                let delim = TokenType::Token(token::CloseDelim(delim));
+
+                // We want to suggest the inclusion of the closing delimiter where it makes
+                // the most sense, which is immediately after the last token:
+                //
+                //  {foo(bar {}}
+                //      -   -  ^ expected one of `)`, <...>
+                //      |   |
+                //      |   help: ...missing `)` might belong here
+                //      you might have meant to close this...
```

Comment:
is this really the right spot? I would have suggested us to suggest adding the `)` after the `{}`, like so `{foo(bar {})}`, not `{foo(bar) {}}`, which doesn't look to me like it would parse

Reply:
Now I'm intrigued on wether the comment is correct, at all...

I'd like to create a bunch of real "broken" code samples to test against, which would give me more confidence that what I'm doing in this PR is beneficial.

Reply:
That would help me too, since I'm starting my reviews by just looking at the test cases (I haven't really looked at the code at all yet)